### PR TITLE
jobs: enhance the InfoStorage API

### PR DESCRIPTION
**Pull request by knz**
Needed for #98459.
Epic: CRDB-23559

### CODE REVIEW

File: pkg/jobs/job_info_storage.go [hunk 1]
```diff
@@ -138,19 +146,32 @@ func (i InfoStorage) write(ctx context.Context, infoKey, value []byte) error {
...
+	fn func(infoKey, value []byte) error,
 ) (retErr error) {
 	if i.txn == nil {
 		return errors.New("cannot iterate over the job info table without an associated txn")
 	}
 
+	var iterConfig string
+	switch iterMode {
+	case iterateAll:
+		iterConfig = `ORDER BY info_key ASC, written DESC` // with no LIMIT.
```
Comment (adityamaru) [inline pkg/jobs/job_info_storage.go:169]: given that for a job_id there will only ever be one row per `info_key` we could drop the secondary sort on written, right?

**Review by adityamaru — APPROVED** [review]
Thanks for doing this, just a couple of non-blocking comments!

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98993)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru)

___
*[`pkg/jobs/job_info_storage.go` line 161 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98993#-NQvrfjMXNTGZw1oBpr4-r3-161:-NR-r2hS6G9XgbBq4qV2:b-9hmdgz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce35a2be1a9247f3391e393f5d59f48bc2e11e8/pkg/jobs/job_info_storage.go#L161)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

given that for a job_id there will only ever be one row per `info_key` we could drop the secondary sort on written, right?
</blockquote></details>

Discussed internally here https://cockroachlabs.slack.com/archives/C01RX2G8LT1/p1679345511855089 - in summary it doesn't matter here.
I'm going to keep it with an explanatory comment.
___
*[`pkg/jobs/job_info_storage_test.go` line 162 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98993#-NQvrfjMXNTGZw1oBpr6-r3-162:-NR-nC7k2iuw8YuLeT5y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce35a2be1a9247f3391e393f5d59f48bc2e11e8/pkg/jobs/job_info_storage_test.go#L162)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I think `kC` only has one revision, could we write a few more revisions before iterating?
</blockquote></details>

Done.

**Comment by knz** [comment]
TFYR!

bors r=adityamaru

**Comment by knz** [comment]
blathers backport 23.1

**Pull request merged**